src/Collection/ArrayCollection.php

```diff
@@ -0,0 +1,247 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Arrays\Collection;
+
+use ArrayAccess;
+use ArrayIterator;
+use Countable;
+use IteratorAggregate;
+use Yiisoft\Arrays\Collection\Modifier\ModifierInterface\AfterMergeModifierInterface;
+use Yiisoft\Arrays\Collection\Modifier\ModifierInterface\BeforeMergeModifierInterface;
+use Yiisoft\Arrays\Collection\Modifier\ModifierInterface\DataModifierInterface;
+use Yiisoft\Arrays\Collection\Modifier\ModifierInterface\ModifierInterface;
+
+final class ArrayCollection implements ArrayAccess, IteratorAggregate, Countable
+{
+    private array $data;
+
+    /**
+     * @var array|null Result array cache
+     */
+    private ?array $array = null;
+
+    /**
+     * @var ModifierInterface[]
+     */
+    private array $modifiers;
+
+    /**
+     * @param array $data
+     * @param ModifierInterface|ModifierInterface[] $modifiers
+     */
+    public function __construct(array $data = [], $modifiers = [])
+    {
+        $this->data = $data;
+        $this->modifiers = is_array($modifiers) ? $modifiers : [$modifiers];
```

Comment:
What do you think about this:

```suggestion
    public function __construct(array $data = [], ModifierInterface ...$modifiers)
    {
        $this->data = $data;
        $this->modifiers = $modifiers;
```

Reply:
```php
$arrayCollection = new ArrayCollection($data, new Modifier1(), new Modifier2());

// vs

$arrayCollection = new ArrayCollection($data, [new Modifier1(), new Modifier2()]);
```

I'm alright with both. @vjik it is you to decide which variant you like better.